pkg/p2p/peer/connector.go

```diff
@@ -0,0 +1,198 @@
+// This Source Code Form is subject to the terms of the MIT License.
+// If a copy of the MIT License was not distributed with this
+// file, you can obtain one at https://opensource.org/licenses/MIT.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+package peer
+
+import (
+	"bytes"
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/dusk-network/dusk-blockchain/pkg/config"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/message"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/protocol"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
+	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/eventbus"
+	log "github.com/sirupsen/logrus"
+)
+
+type connectFunc func(context.Context, *Reader, *Writer, chan bytes.Buffer)
+
+// Connector is responsible for accepting incoming connection requests, and
+// establishing outward connections with desired peers.
+type Connector struct {
+	eventBus      eventbus.Broker
+	gossip        *protocol.Gossip
+	readerFactory *ReaderFactory
+
+	l net.Listener
+
+	lock     sync.RWMutex
+	registry map[string]struct{}
+
+	services protocol.ServiceFlag
+
+	connectFunc connectFunc
+}
+
+// NewConnector creates a new peer connector, and spawns a goroutine that will
+// accept incoming connection requests on the current address, with the given port.
+func NewConnector(eb eventbus.Broker, gossip *protocol.Gossip, port string,
+	processor *MessageProcessor, services protocol.ServiceFlag,
+	connectFunc connectFunc) *Connector {
+	addrPort := ":" + port
+
+	listener, err := net.Listen("tcp", addrPort)
+	if err != nil {
+		log.WithField("process", "peer connector").
+			WithError(err).
+			Panic("could not establish a listener")
+	}
+
+	c := &Connector{
+		eventBus:      eb,
+		gossip:        gossip,
+		readerFactory: NewReaderFactory(processor),
+		l:             listener,
+		registry:      make(map[string]struct{}),
+		services:      services,
+		connectFunc:   connectFunc,
+	}
+
+	processor.Register(topics.Addr, c.ProcessNewAddress)
+
+	go func(c *Connector) {
+		for {
+			conn, err := c.l.Accept()
+			if err != nil {
+				log.WithField("process", "peer connector").
+					WithError(err).
+					Warnln("error accepting connection request")
+				return
+			}
+
+			go c.acceptConnection(conn)
+		}
+	}(c)
+
+	return c
+}
+
+// Close the listener.
+func (c *Connector) Close() error {
+	return c.l.Close()
+}
+
+// ProcessNewAddress will handle a new Addr message from the network.
+// Satisfies the peer.ProcessorFunc interface.
+func (c *Connector) ProcessNewAddress(srcPeerID string, m message.Message) ([]bytes.Buffer, error) {
+	a := m.Payload().(message.Addr)
+	return nil, c.Connect(a.NetAddr)
```

Comment:
An OS usually has fd max limit, if a node is flooded with `topics.Addr`, this limit could be reached and the goroutine would panic. Shall we limit here the number of dialing a connection?